src/libsystemd/sd-device/device-enumerator.c

```diff
@@ -541,29 +539,70 @@ static int match_initialized(sd_device_enumerator *enumerator, sd_device *device
         return (enumerator->match_initialized == MATCH_INITIALIZED_NO) == (r == 0);
 }
 
+static bool match_subsystem(sd_device_enumerator *enumerator, const char *subsystem) {
+        assert(enumerator);
+
+        if (!subsystem)
+                return false;
+
+        return set_fnmatch(enumerator->match_subsystem, enumerator->nomatch_subsystem, subsystem);
+}
+
+typedef enum MatchFlag {
+        MATCH_SYSNAME     = 1u << 0,
+        MATCH_SUBSYSTEM   = 1u << 1,
+        MATCH_PARENT      = 1u << 2,
+        MATCH_TAG         = 1u << 3,
+
+        MATCH_ALL         = (1u << 4) - 1,
+} MatchFlag;
+
 static int test_matches(
                 sd_device_enumerator *enumerator,
                 sd_device *device,
-                bool ignore_parent_match) {
+                MatchFlag flags) {
 
         int r;
 
         assert(enumerator);
         assert(device);
 
-        /* Checks all matches, except for the sysname match (which the caller should check beforehand) */
+        if (FLAGS_SET(flags, MATCH_SYSNAME)) {
+                const char *sysname;
 
-        r = match_initialized(enumerator, device);
-        if (r <= 0)
-                return r;
+                r = sd_device_get_sysname(device, &sysname);
+                if (r < 0)
+                        return r;
```

Comment:
Can we move this into `match_sysname()`? Ditto for subsystem, that way all match helper functions have the same API

Reply:
Unfortunately no, as match_sysname() and match_subsystem() are called with name `struct dirent`.